libstuff/libstuff.cpp

```diff
@@ -1899,23 +1914,29 @@ bool S_sendconsume(int s, string& sendBuffer) {
     chrono::steady_clock::time_point start = chrono::steady_clock::now();
 
     // Send as much as we can
-    ssize_t numSent = send(s, sendBuffer.c_str(), (int)sendBuffer.size(), MSG_NOSIGNAL);
+    ssize_t numSent = send(s, sendBuffer.c_str(), sendBuffer.size(), MSG_NOSIGNAL);
     string errorMessage;
     if (numSent == -1) {
         errorMessage = " Error: "s + strerror(errno);
     }
     SINFO("[performance] Send() took " << chrono::duration_cast<chrono::milliseconds>(chrono::steady_clock::now() - start).count()
-        << " ms and sent " << numSent << " of " << (int)sendBuffer.size() << " bytes." << errorMessage);
+        << " ms and sent " << numSent << " of " << sendBuffer.size() << " bytes." << errorMessage);
 
     if (numSent > 0) {
         SConsumeFront(sendBuffer, numSent);
     }
 
-    // Exit of no error
+    // Exit if no error
     if (numSent >= 0) {
         return true; // No error; still alive
     }
 
+    // EAGAIN/EWOULDBLOCK are fatal for send cases
```

Comment:
Why are these fatal?

Reply:
They really shouldn't be, and this will probably get reverted once we figure out how we get so far ahead of certain followers, but the intention was to reconnect rather than crash once our send buffer was full, and so this is a bit of a stopgap measure for that i n the meantime.

Reply:
If we get EAGAN on a send() call, it means the local send buffer is full, which means the socket is non-functioning. If we treat the socket as still connected, we'll append to the internal (bedrock) buffer for this socket indefinitely until we can eventually call send() successfully again which is what caused the crash last week. If we treat send() EAGAINs as fatal and disconnect the socket, the reconnect time may be enough to clear the send() socket and break us out of the indefinite backfill loop. 

Most of this PR is experimental and an attempt to (1) prevent us from crashing again tomorrow night and (2) get us more information about how the neighbor peer is getting so far behind when the remote peers are not.

Reply:
Hm, but if the goal is to simply "not crash", how does this line help?  The crash line happened from an int overflow thing triggering an assert -- this seems totally unrelated to that.  My fear is we might get EAGAIN in a million scenarios other than the one we care about here, so it could cause reconnects totally unnecessarily.  Especially in a scenario where we are like fighting some kind of networking problem in realtime that triggers this landmine.

Just saying, if we don't intend to keep it, let's not put it in, because we are absolutely going to forget about it until it explodes.

Reply:
I think that concern is valid, but I don't think we have any other good ideas for "What do we do when we keep trying to send to someone and it never works?"

Like I said, this is a stopgap, and we could handle this some other way than watching for EAGAIN on `send`, but what is that? When do we give up on sending to someone? Last time, we gave up by crashing when we'd buffered 2GB of data to send. Next time we won't overflow in this case, but we'll still give up by crashing when we segfault from running out of memory.

@coleaeason how long did it take to go from the first `send` that didn't send all of it's data to the crash with a 2GB buffer? We can imagine this time could be up to 325x as long before running out of memory on one of our production servers.

We could just reconnect if the send buffer grows past some magic number instead, but what we don't want to do is let it grow forever.

We don't know of any other case where this has been observed to happen (`EAGAIN` on `send()`, which is why this seems reasonable to try, but you're right that we may run into some other error condition where this occurs and causes an infinite re-connection loop.

Reply:
If our concern is the size of the send buffer, let's reconnect on that.  I think it's entirely possible to get EAGAIN on some other benign, transient condition.

Reply:
It took almost exactly 12 hours to get to 2gb.

However letting it grow forever is still not good enough on the premise of "it takes a long longer on our giant servers". Bedrock should be able to run on hardware lesser than our monster production machines, so I don't like the justification of "now we can wait longer for it to fail" because someone else running this on a reasonable machine with 4gb of RAM can't.

Reply:
I agree we should do something, I just think this is a very untargeted solution to what is actually a very narrow problem.  If we are concerned by the size of the buffer, let's reconnect when we go over a buffer size.  If we are concerned that we haven't been able to send for 12 hours (which seems like a different problem entirely), let's track how long between sends and reconnect on that.  I just think EAGAIN is not the right way to solve this problem.

Reply:
I'm fine with making one or the other of the changes @quinthar suggests here instead. Reconnecting at a certain buffer size is easier to implement, which is probably fine for this testing and diagnosis that we're trying to do. Set a limit of 1gb and check for the size of the buffer after failing to send, and reconnect if the buffer is over the limit.

Reply:
Modified here: https://github.com/Expensify/Bedrock/pull/651

Errors will be fatal and cause the socket to disconnect if they occur with over 1GB in the send buffer, even if the4y would normally be non-fatal.